### DISCUSSION TITLE
Csla.DataPortalException.BusinessException in 5.3.0

### DISCUSSION THREAD

**Discussion by ajohnstone-ks** [MarimerLLC/csla#1763] [Questions] [answered]
We have some tests which are failing now, and I tracked it down to the fact that Csla.DataPortalException.BusinessException is returning a Csla.Reflection.CallMethodException, which has its inner exception as the one actually thrown from within the Child_Update exception.  

Is this change intentional?  Previously it was the exception directly throw from within the Child_Update method.

**Comment by rockfordlhotka** [comment, marked as answer]
Interesting. Not intentional.

The way the data portal invokes methods is entirely different in version 5, because we added the ability to inject parameter values and support for the data portal operation attributes (`Create`, `Fetch`, `Update`, etc.).

The [`BusinessException` method](https://github.com/MarimerLLC/csla/blob/master/Source/Csla.Shared/DataPortalException.cs#L167) tries to strip out `DataPortalException` and `CallMethodException` exceptions to find the underlying exception. If you look at the code it is pretty simplistic, and perhaps there's some intermediate exception in the way?

Can you provide the entire stack trace (exception tree) from your test?

**Reply by ajohnstone-ks** [reply]
It almost looks like an off by 1 error.  Is this what you're looking for?  The exception highlighted in blue is our actual exception which were were expecting in BusinessException.

![CallMethodException](https://user-images.githubusercontent.com/22680122/90153241-b877b400-dd56-11ea-8c21-2450338eccff.png)

**Reply by rockfordlhotka** [reply]
This is what I was looking for, thank you.

The implementation of `BusinessException` has never been robust, and that is causing the issue.

Though I am also wondering why there are 3 `CallMethodException` instances in the stack, that seems odd to me. It makes me wonder if there's something not working right in the way the methods are being invoked, or the way exceptions are being thrown as part of the invocation process.

**Reply by rockfordlhotka** [reply]
I've recorded this as a bug: https://github.com/MarimerLLC/csla/issues/1764

**Reply by ajohnstone-ks** [reply]
I doubt it matters, but this is via the WCF data portal.  

I think I've seen multiple nested exceptions like this before, I think because the object throwing the exception is actually a grand-child object, so it seems like each level is adding its own CallMethodException.

**Comment by rockfordlhotka** [comment]
@ajohnstone-ks does this look like a workable solution to you?

https://github.com/MarimerLLC/csla/pull/1776/commits/c1e744d074ba09db2e165fc788e8b567bd98856a
